src/adapters/nats.js

```diff
@@ -0,0 +1,436 @@
+						ack_policy: "explicit",
+						// More info: https://docs.nats.io/jetstream/concepts/consumers#maxackpending
+						max_ack_pending: this.opts.maxInFlight
+					}
+				}
+			}
+		});
+
+		// Adapted from: https://github.com/moleculerjs/moleculer/blob/3f7e712a8ce31087c7d333ad9dbaf63617c8497b/src/transporters/nats.js#L141-L143
+		if (this.opts.url)
+			this.opts.nats.connectionOpts.servers = this.opts.url
+				.split(",")
+				.map(server => new URL(server).host);
+
+		/** @type {NatsConnection} */
+		this.connection = null;
+
+		/** @type {JetStreamManager} */
+		this.manager = null;
+
+		/** @type {JetStreamClient} */
+		this.client = null;
+
+		/** @type {Map<string,JetStreamSubscription>} */
+		this.subscriptions = new Map();
+	}
+
+	/**
+	 * Initialize the adapter.
+	 *
+	 * @param {ServiceBroker} broker
+	 * @param {Logger} logger
+	 */
+	init(broker, logger) {
+		super.init(broker, logger);
+
+		try {
+			NATS = require("nats");
+		} catch (err) {
+			/* istanbul ignore next */
+			this.broker.fatal(
+				"The 'nats' package is missing! Please install it with 'npm install nats --save' command.",
+				err,
+				true
+			);
+		}
+
+		this.checkClientLibVersion("nats", "^2.2.0");
+	}
+
+	/**
+	 * Connect to the adapter.
+	 */
+	async connect() {
+		this.connection = await NATS.connect();
+
+		this.manager = await this.connection.jetstreamManager();
+
+		this.client = this.connection.jetstream(); // JetStreamOptions
+	}
+
+	/**
+	 * Disconnect from adapter
+	 */
+	async disconnect() {
+		this.stopping = true;
+
+		try {
+			if (this.connection) {
+				this.logger.info("Closing NATS JetStream connection...");
+				await this.connection.drain();
+				await this.connection.close();
+
+				this.logger.info("NATS JetStream connection closed.");
+			}
+		} catch (error) {
+			this.logger.error("Error while closing NATS JetStream connection.", error);
+		}
+	}
+
+	/**
+	 * Subscribe to a channel with a handler.
+	 *
+	 * @param {Channel & NATSOpts} chan
+	 */
+	async subscribe(chan) {
+		this.logger.info(`NATS --- ${chan.id} --- in <=`);
+
+		if (chan.maxInFlight == null) chan.maxInFlight = this.opts.maxInFlight;
+		if (chan.maxRetries == null) chan.maxRetries = this.opts.maxRetries;
+
+		chan.deadLettering = _.defaultsDeep({}, chan.deadLettering, this.opts.deadLettering);
+		if (chan.deadLettering.enabled) {
+			chan.deadLettering.queueName = this.addPrefixTopic(chan.deadLettering.queueName);
+		}
+
+		// 1. Create stream
+		// NATS Stream name does not support: spaces, tabs, period (.), greater than (>) or asterisk (*) are prohibited.
+		// More info: https://docs.nats.io/jetstream/administration/naming
+		const streamName = chan.name.split(".").join("_");
+		this.createStream(streamName, [chan.name], chan.nats ? chan.nats.streamConfig : {});
+
+		if (chan.deadLettering && chan.deadLettering.enabled) {
+			const deadLetteringStreamName = chan.deadLettering.queueName.split(".").join("_");
+			this.createStream(
```

Comment:
Why should create a stream for dead-lettering? We can't publish message to non-existing topics?

Reply:
To store the message it's necessary to create a stream and define how messages are stored and how long they persist.
Otherwise it throws an error.